src/mantine-core/src/TransferList/TransferList.tsx

```diff
@@ -124,8 +129,19 @@ export const TransferList = forwardRef<HTMLDivElement, TransferListProps>((props
   const handleMoveAll = (listIndex: 0 | 1) => {
     const items: TransferListData = Array(2) as any;
     const moveToIndex = listIndex === 0 ? 1 : 0;
-    items[listIndex] = [];
-    items[moveToIndex] = [...value[moveToIndex], ...value[listIndex]];
+
+    if (transferAllMatchingFilter) {
+      const query = search[listIndex];
+      const shownItems = value[listIndex].filter((item) => filter(query, item)).slice(0, limit);
```

Comment:
Not too sure whether `.slice`-ing here makes sense. E.g: would we want this to only transfer the items that are actually **shown**, or all the items that match the filter?